sdk/core/core/inc/az_json.h

```diff
@@ -128,71 +180,183 @@ typedef struct
   } _internal;
 } az_json_builder;
 
-AZ_NODISCARD AZ_INLINE az_result az_json_builder_init(az_json_builder* self, az_span json_buffer)
+/*
+ * @brief az_json_builder_init initializes an az_json_builder which writes JSON into a buffer.
+ *
+ * @param json_builder A pointer to an az_json_builder instance to initialize.
+ * @param json_buffer An az_span over the buffer where th JSON object is to be written.
+ * @return AZ_OK if the az_json_builder is initialized correctly.
+ */
+AZ_NODISCARD AZ_INLINE az_result
+az_json_builder_init(az_json_builder* json_builder, az_span json_buffer)
 {
-  *self = (az_json_builder){ ._internal = { .json = json_buffer, .need_comma = false } };
+  *json_builder = (az_json_builder){ ._internal = { .json = json_buffer, .need_comma = false } };
   return AZ_OK;
 }
 
-AZ_NODISCARD AZ_INLINE az_span az_json_builder_span_get(az_json_builder const* self)
+/*
+ * @brief az_json_builder_span_get returns the az_span containing the final JSON object.
+ *
+ * @param json_builder A pointer to an az_json_builder instance wrapping the JSON buffer.
+ * @return an az_span containing the final JSON object.
+ */
+AZ_NODISCARD AZ_INLINE az_span az_json_builder_span_get(az_json_builder const* json_builder)
 {
-  return self->_internal.json;
+  return json_builder->_internal.json;
 }
 
-AZ_NODISCARD az_result az_json_builder_append_token(az_json_builder* self, az_json_token token);
-
+/*
+ * @brief az_json_builder_append_token appends an az_json_token to the JSON buffer.
+ *
+ * @param json_builder A pointer to an az_json_builder instance containing the buffer to append the
+ * token to.
+ * @param token A pointer to the az_json_token to append.
+ * @return AZ_OK if the token was appended successfully.<br>
+ * AZ_ERROR_BUFFER_OVERFLOW if the buffer is too small.
+ */
 AZ_NODISCARD az_result
-az_json_builder_append_object(az_json_builder* self, az_span name, az_json_token token);
-
-AZ_NODISCARD az_result az_json_builder_append_object_close(az_json_builder* self);
+az_json_builder_append_token(az_json_builder* json_builder, az_json_token token);
 
+/*
+ * @brief az_json_builder_append_object appends a JSON to the JSON buffer.
+ *
+ * @param json_builder A pointer to an az_json_builder instance containing the buffer to append the
+ * object to.
+ * @param name A span containing the JSON object's name.
+ * @param token A pointer to the az_json_token to append as the object's value.
+ * @return AZ_OK if the token was appended successfully.<br>
+ * AZ_ERROR_BUFFER_OVERFLOW if the buffer is too small.
+ */
 AZ_NODISCARD az_result
-az_json_builder_append_array_item(az_json_builder* self, az_json_token token);
+az_json_builder_append_object(az_json_builder* json_builder, az_span name, az_json_token token);
```

Comment:
What happens if someone calls this method when inside a JSON array (which could result in invalid JSON)?
```json
[ 1, 2, "foo":{}, 4]
```

Reply:
Well, it could be an array of arrays in which case it would be valid JSON.
We can't protect against this. Someone can always just write a string with invalid JSON and not use our APIs at all.

Reply:
Of course, but since I am using a JSON builder, shouldn't that help validate such things? Otherwise, we are reducing part of the benefit of such an API, no?

Reply:
How do we validate if an array has array elements to allow or disallow this?

Reply:
Based on the current state of the builder. If the current state/token of the builder is "writing inside an array", then any calls to `az_json_builder_append_object` should fail with an error since writing a property name within an array is invalid JSON.

Similarly, if the current state is "writing inside an object", calling anything that doesn't write a a propertyname first show fail.

> How do we validate if an array has array elements to allow or disallow this?

I don't see how arrays of arrays is related here. Also I don't see how whether an array has a single element, or is empty, or has array elements in it, is related to validation. You only need to know about the state of the inner most container (if it is array or object).

To do this, you would need some sort of stack as part of the `az_json_builder` state.

**So, here are a couple of invalid cases:**
- writing a single JSON value inside a JSON object without accompanying key name
```json
{"just a value with no key"}
```
- writing a key-value pair inside a JSON array instead of just a value
```json
["key":"value ]
```

Reply:
@ahsonkhan , calling `az_json_builder_append_object()` inside an array would add brackets first and would end up as
```json
{ "array": [1, 2, { "obj": value}, 3] }
``` 
I will add an extra test for this so you can see it.
There is no a way of adding `:` directly with our parser.

The only way we can end up with a bad Json is by using append_array_item() without appending an array_start. This would generate some invalid thing like
```json
{ array_element }
```
Or calling append_token directly without making sure object or array was started.

However, if we add validation for this, we would be adding more code and size to something that looks like as a nice-to-have thing

Reply:
> calling `az_json_builder_append_object()` inside an array would add brackets first

I might be reading the code wrong, but that doesn't look like its true.

This test is passing and producing invalid JSON.

```c
uint8_t array1[200];
az_json_builder builder1 = { 0 };

TEST_EXPECT_SUCCESS(az_json_builder_init(&builder1, AZ_SPAN_FROM_BUFFER(array1)));

TEST_EXPECT_SUCCESS(az_json_builder_append_token(&builder1, az_json_token_array()));

TEST_EXPECT_SUCCESS(az_json_builder_append_object(
    &builder1, AZ_SPAN_FROM_STR("int-max"), az_json_token_number(9007199254740991ull)));

TEST_EXPECT_SUCCESS(az_json_builder_append_array_close(&builder1));

assert_true(az_span_is_content_equal(
    builder1._internal.json,
    AZ_SPAN_FROM_STR( //
        "["
        "\"int-max\":9007199254740991"
        "]")));
```